[RUN-4040] Use Execution API in Activity Page instead EventAjax

ltamaster opened the pull request:
## Tell us about your PR

**Is this a bugfix, or an enhancement? Please describe.**

Replace the use of eventAjax (Dynamic queries for BaseReport table) with the use of the Execution API for the Activity Page. This provides better performance, cleaner architecture, and improved scalability for installations with large execution tables.

For the execution API, it needs some changes to cover the Activity Page functionality:
- Improve the criteria builder with better filters and optimized sorting
- Add optimized HQL count for cross-project job references (includeJobRef)
- Add simple HQL count without JOINs (configurable via `countPerformance.enabled`)
- Add execution count caching

---

## Summary of Changes

### Components Modified

| Layer | Files | Changes |
|-------|-------|---------|
| **Frontend** | `activityList.vue`, `executions.ts` | Use Execution API instead of eventAjax |
| **API** | `ExecutionController.groovy` | New filters, `olderFilter` support, partial match descriptions |
| **Backend** | `ExecutionService.groovy` | Criteria builder improvements, simple HQL count, count caching |
| **Query** | `ExecutionQuery.groovy` | New filters, HQL count methods, simple count, LIKE escaping |
| **Domain** | `Execution.groovy` | Removed hasOne relationship for performance |
| **Tests** | `ExecutionServiceExecutionApiIntegrationSpec.groovy` | Comprehensive filter and count tests |

---

## Feature 1: Improved Criteria Builder

Enhancements to the execution query criteria builder for better performance and functionality.

**Goals:**
- **Remove extra queries**: Eliminate unnecessary database queries triggered by `executionSummary` (workflow steps) and `LogFileStorageRequest` (hasOne relationship)
- **Better index usage**: Optimize sorting to leverage database indexes (using `id` instead of `dateStarted` for secondary sort)
- **Legacy status compatibility**: Support both legacy (`'false'`, `'true'`) and new (`'failed'`, `'succeeded'`) status values in filters

### 1.1 Optimized Sort Fields

Changed the default sorting from `dateCompleted DESC, dateStarted DESC` to `dateCompleted DESC, id DESC`:

```groovy
order('dateCompleted', 'desc')
order('id', 'desc')  // Primary key - always indexed, faster than dateStarted
```

**Why `id` instead of `dateStarted`?**
- **Indexed**: Primary key is always indexed
- **Sequential**: IDs are assigned in creation order
- **Unique**: Guarantees deterministic ordering for pagination
- **Fast**: No additional index lookup required

### 1.2 New Query Filters

| Filter | Type | Description |
|--------|------|-------------|
| `olderFilter` | String | Relative date (e.g., "60d", "24h") - executions older than |
| `recentFilter` | String | Relative date for recent executions |
| `nodeFilter` | String | Filter by node name (partial match with ILIKE) |
| `optionFilter` | String | Filter by option values (partial match with ILIKE) |
| `adhocStringFilter` | String | Filter adhoc commands by title (partial match) |
| `excludeRunning` | Boolean | Exclude running executions (`dateCompleted IS NOT NULL`) |
| `executionSummary` | Boolean | Include/exclude summary field in response |

### 1.3 LIKE Pattern Escaping (Security)

User input in `nodeFilter`, `optionFilter`, and `adhocStringFilter` is sanitized to prevent SQL wildcard injection.

### 1.4 Disabled executionSummary by Default

The `executionSummary` field is now **disabled by default** in the JSON response, avoiding extra queries to workflow tables.

### 1.5 Domain Model Optimization

Removed `hasOne` relationship between `Execution` and `LogFileStorageRequest` to prevent unwanted JOINs during list queries.

### 1.6 Legacy Status Value Support

The status filter now correctly matches both legacy and new status values:

| Status Filter | Matches |
|--------------|---------|
| `failed` | `'failed'` OR `'false'` (legacy) |
| `succeeded` | `'succeeded'` OR `'true'` (legacy) |

---

## Feature 2: Cross-Project Job Reference Count Optimization

Optimized HQL-based count queries for `includeJobRef` queries that count executions across multiple projects.

### Optimized UNION Query for includeJobRef

For queries with `includeJobRef` (cross-project job references), uses a dual-query approach:

```sql
-- COUNT QUERY 1: Referenced executions (via workflow job reference steps)
SELECT COUNT(e.id)
FROM Execution e
INNER JOIN ReferencedExecution re WITH re.execution.id = e.id
WHERE e.scheduledExecution IS NOT NULL
AND re.jobUuid = :jobUuid
AND e.project IN (:projects)

-- COUNT QUERY 2: Direct executions (without ReferencedExecution)
SELECT COUNT(e.id)
FROM Execution e
WHERE e.project = :project
AND e.scheduledExecution IS NOT NULL
AND e.jobUuid = :jobUuid
AND NOT EXISTS (SELECT 1 FROM ReferencedExecution re WHERE re.execution.id = e.id)
```

**Performance improvement**: 50-100x faster than EXISTS-based approach using INNER JOIN.

### When is this used?

The HQL count optimization is used when `shouldUseUnionQuery()` returns true:
- `includeJobRef` is enabled
- A single job UUID filter is specified
- Both `projFilter` and `execProjects` are set

---

## Feature 3: Simple HQL Count (Configurable)

Optimized HQL count query that avoids expensive JOINs with `scheduled_execution` table.

### The Problem

The default criteria-based count generates a JOIN query:

```sql
-- Default criteria count (slower under heavy load)
SELECT count(*)
FROM execution this_
INNER JOIN scheduled_execution schedulede1_
    ON this_.scheduled_execution_id = schedulede1_.id
WHERE this_.scheduled_execution_id IS NOT NULL
  AND schedulede1_.uuid = 'job-uuid'
  AND schedulede1_.project = 'MyProject'
  AND this_.project = 'MyProject'
  AND date_completed BETWEEN '2025-01-16' AND '2026-01-21';
```

### The Solution

Use a simple HQL query that filters directly on `execution.job_uuid`:

```sql
-- Simple HQL count (faster, no JOIN)
SELECT count(execution0_.id)
FROM execution execution0_
WHERE execution0_.project = 'MyProject'
  AND execution0_.date_completed BETWEEN '2025-01-16' AND '2026-01-21'
  AND execution0_.job_uuid = 'job-uuid';
```

### Configuration

```properties
# Enable simple HQL count optimization (default: false)
rundeck.api.executionQueryConfig.countPerformance.enabled=true
```

| Setting | Behavior |
|---------|----------|
| `countPerformance.enabled=true` | Uses simple HQL count (no JOINs) |
| `countPerformance.enabled=false` | Uses criteria-based count (default) |

### Required Index

This optimization requires a new composite index:

```sql
CREATE INDEX EXEC_IDX_JOB_PROJ_DATE
ON execution (job_uuid, project, date_completed, status, cancelled);
```

**Note**: This index is created in a separate PR ([#9964](https://github.com/rundeck/rundeck/pull/9964)).

### When is Simple Count Used?

The `canUseSimpleCount()` method returns `true` when:
- `countPerformance.enabled=true`
- No ILIKE filters (`nodeFilter`, `optionFilter`, `adhocStringFilter`)
- Uses job UUID filters (not numeric IDs)

### Performance Comparison

| Query Type | Time (11M rows) | Notes |
|------------|-----------------|-------|
| Criteria with JOIN | 7-50 seconds | Under heavy load |
| Simple HQL (no JOIN) | &lt;1 second | With index |

---

## Feature 4: Execution Count Cache

Caches the results of expensive `COUNT(*)` queries to avoid repeated database scans.

### Configuration

```properties
# Enable/disable count cache (default: false)
rundeck.api.executionQueryConfig.countCache.enabled=true

# Cache TTL in seconds (default: 30)
rundeck.api.executionQueryConfig.countCache.ttl=30
```

| Setting | Behavior |
|---------|----------|
| `countCache.enabled=true` | Caches COUNT query results |
| `countCache.enabled=false` | Disables caching, runs COUNT every time (default) |

### Features

- **Thread-Safe Cache Initialization**: Uses double-checked locking with volatile field
- **Stale Cache Detection**: Automatically detects and refreshes stale cached counts
- **Cache Key Structure**: Type-safe immutable cache keys with static invalidation methods

---

## Performance Impact

| Metric | Before | After |
|--------|--------|-------|
| Activity page load | 7-50 seconds | &lt;2 seconds |
| COUNT query time | 7-50 seconds | &lt;100ms (cache hit) or &lt;1s (simple HQL) |
| MySQL connections | Exhausted under load | Stable |

---

## Testing

- **Unit Tests**: `ExecutionServiceTests.groovy`, `ExecutionQuerySpec.groovy` - Cache key generation, cacheability, LIKE escaping, canUseSimpleCount
- **Integration Tests**: `ExecutionServiceExecutionApiIntegrationSpec.groovy` - Cache behavior, filter logic, count queries
- **Functional Tests**: `JobActivityHistorySpec.groovy` - Activity page functionality

---

## Limitations

1. **Multiple Job IDs**: Queries with multiple job IDs are not cached
2. **Absolute Dates**: Queries with absolute date filters are not cached
3. **Start Date Filters**: `startafterFilter`, `startbeforeFilter` not cached
4. **Stale Counts**: Counts may be up to TTL seconds stale (when cache is enabled)
5. **ILIKE Filters**: Simple HQL count cannot be used when ILIKE filters are present

---

## Release Notes

Performance optimization for Activity Page: Replaced eventAjax with Execution API. Includes four major improvements:
1. **Criteria Builder**: Optimized sorting (`id` instead of `dateStarted`), new filters, LIKE escaping for security, legacy status value support (`'false'`/`'true'`)
2. **Cross-Project Job Reference Count**: HQL-based count queries for `includeJobRef` avoiding expensive EXISTS subqueries
3. **Simple HQL Count**: Configurable HQL count without JOINs for job-filtered queries (requires index, configurable via `countPerformance.enabled`)
4. **Count Cache**: Caches COUNT results to reduce database load (configurable via `countCache.enabled`)

Reduces database load and improves response times for large execution tables.

ltamaster committed:
improve response time of execution API calls

ltamaster committed:
remove descending in index

ltamaster committed:
fix test

ltamaster committed:
use execution API in activity page

ltamaster committed:
set a new index

ltamaster committed:
add the other filters by option, node filter, node, adhoc command

ltamaster committed:
fix custom status

ltamaster committed:
fix gui unit test
split the count query with a jobReference filter in execution API for performace issues

ltamaster committed:
remove project filter is not applied when the jobUUID filter is set to improve the response time of that case

ltamaster committed:
use like instead ilike

ltamaster committed:
new index that improve response time of execution API with filters

ltamaster committed:
set filter by dateCompleted to improve performance when only project filter is set

ltamaster committed:
check if lastExecutionId is not used

ltamaster committed:
use waitForExecutionFinish in tests

ltamaster committed:
remove unnecessary test
fix command page

ltamaster committed:
remove default filter by dateCompleted when the project is set

ltamaster committed:
simplify executeJobReferenceCount
fix text activityList.spec.ts
enable the code that detect a deleted job in activity page
try to remove unnecessary calls to Execution.get(id) in ApiService

ltamaster committed:
fix unit test

ltamaster committed:
use group+name in activity page
remove index creation

ltamaster committed:
fix job name in activity list

ltamaster committed:
remove date complete form status

ltamaster committed:
improve performance of query with status

ltamaster committed:
fix issue with date-complete filter

ltamaster committed:
remove filter by date_completed with project because it breaks a functionality

ltamaster committed:
add a filter to exclude running executions

ronaveva commented on functional-test/src/test/groovy/org/rundeck/tests/functional/selenium/jobs/JobActivityHistorySpec.groovy:
```diff
@@ -45,23 +45,6 @@ class JobActivityHistorySpec extends SeleniumBase {
         }
     }
 
-    def "review job activity history from Job List Page Saved Filters"() {
```
why is this test being deleted?

ronaveva commented on rundeckapp/grails-app/domain/rundeck/Execution.groovy:
```diff
@@ -186,6 +186,13 @@ class Execution extends ExecutionContext implements EmbeddedJsonData, ExecutionD
         return logFileStorageRequest?.id
     }
 
+    Boolean getJobDeleted() {
```
since this returns a boolean a more appropriate name would be "isJobDeleted"

ronaveva commented on rundeckapp/grails-spa/packages/ui-trellis/src/library/services/executions.ts:
```diff
@@ -9,14 +9,16 @@ export interface PagedResult<T> {
   max: number;
   offset: number;
+  total?: number;
+  count?: number;
```
whats the difference between total and count?

ronaveva commented on rundeckapp/src/main/groovy/com/dtolabs/rundeck/app/support/ExecutionQuery.groovy:
```diff
@@ -405,22 +421,14 @@ class ExecutionQuery extends ScheduledExecutionQuery implements Validateable{
-        }else if(state == ExecutionService.EXECUTION_SUCCEEDED){
           isNotNull('dateCompleted')
+        }else if(state == ExecutionService.EXECUTION_SUCCEEDED){
+          eq('status',  'succeeded')
```
the status "true" is no longer considered, is this an old behavior?

ronaveva reviewed (dismissed):
Hello @ltamaster theres a couple small obs to take a look at

ltamaster commented on rundeckapp/src/main/groovy/com/dtolabs/rundeck/app/support/ExecutionQuery.groovy:
```diff
@@ -291,7 +302,7 @@ class ExecutionQuery extends ScheduledExecutionQuery implements Validateable{
             txtfilters.each { key, val ->
               if (query["${key}Filter"]) {
-                ilike(val, '%' + query["${key}Filter"] + '%')
+                like(val, '%' + query["${key}Filter"] + '%')
```
I will return to the previous version

ltamaster committed:
back to ilike in execution query
improve the separated count for job reference
clean commeted code

ltamaster committed:
add new index on execution  table

ltamaster committed:
clean

ltamaster committed:
back to like in group filter

ltamaster committed:
clean status filter

ltamaster committed:
put back the project filter when jobfilter is set

ltamaster committed:
add a executionSummary field in ExecutionQuery to enable/disable execution summary in the execution API report

ltamaster committed:
improve transaction in queryExecutions

ltamaster committed:
fix test ExecutionControllerSpec.groovy

ltamaster committed:
try lazy in logFileStorageRequest for execution table

ltamaster committed:
fix removed class

ltamaster committed:
rollback lazy logFileStorageRequest change

ltamaster committed:
remove transaction

ltamaster committed:
try again logFileStorageRequest lazy

ltamaster committed:
use projection to avoid query to logFileStorageRequest

ltamaster committed:
remove Execution reference to logFileStorageRequest

ltamaster committed:
add cache for count totals of execution API (with a flag)
simplify count query with HQL queries to avoid joins
add functional test for new changes
add tests

ltamaster committed:
remove join in count query for job

ltamaster committed:
use count(*) instead count(id)

smartinellibenedetti commented on rundeckapp/grails-app/services/rundeck/services/ExecutionService.groovy:
```diff
@@ -4107,15 +4148,289 @@ class ExecutionService implements ApplicationContextAware, StepExecutor, NodeSte
+        if (isExecutionCountCacheEnabled() && cacheKey) {
+            getExecutionCountCache().put(cacheKey, total as Long)
+        }
+    }
```
question: I saw that the executions endpoint is being called with excludeRunning = true from the activityPage, but it seems like neither countDirectAndReferencedExecutions nor countExecutionsSimple is including a query to check that dateCompleted is not null. Is this intentional?

smartinellibenedetti reviewed (approved):
There's just this comment that I think it would be good to take a look/address if there's time before merging: https://github.com/rundeck/rundeck/pull/9905#discussion_r2761342033

But leaving the PR approved in case that comment isn't applicable and you'd like to merge as is.

ltamaster committed:
improve the cache, we can use cache in cases we don't have multiples jobs or fixed dates
fix some comments from pair review

ltamaster committed:
worng check to build canUseSimpleCount()

ltamaster committed:
fix tests

ltamaster committed:
fix copilot comments

ltamaster committed:
fix copilot comments v2

ltamaster committed:
fix copilot comments v3: add the ID for sorting
add a feature flag for the performance count mechanims

ltamaster committed:
fix feature flag CountPerformance

ltamaster committed:
fix test for feature flag CountPerformance

ltamaster committed:
remove the custom count query because it needs a second index.

ltamaster committed:
try again with simple count

ltamaster committed:
add feature flag for count performance

ltamaster committed:
Merge branch 'main' into RUN-3906

ltamaster committed:
Merge branch 'main' into RUN-3906

ltamaster committed:
Merge branch 'main' into RUN-3906

ltamaster committed:
Merge branch 'main' into RUN-3906

ltamaster commented:
@copilot resolve the merge conflicts in this pull request

Copilot committed:
Merge remote-tracking branch 'origin/main' into RUN-3906

# Conflicts:
#	rundeckapp/grails-app/services/rundeck/services/ApiService.groovy
#	rundeckapp/src/test/groovy/rundeck/controllers/ScheduledExecutionControllerSpec.groovy

Co-authored-by: ltamaster <6034968+ltamaster@users.noreply.github.com>

ltamaster commented:
@copilot resolve the merge conflicts in this pull request

Copilot committed:
Merge remote-tracking branch 'origin/main' into RUN-3906

# Conflicts:
#	rundeckapp/grails-spa/packages/ui-trellis/src/app/components/activity/tests/activityList.spec.ts

Co-authored-by: ltamaster <6034968+ltamaster@users.noreply.github.com>

ltamaster committed:
Merge branch 'main' into RUN-3906
